go/vt/vtgateproxy/discovery.go

```diff
@@ -96,60 +91,87 @@ func RegisterJsonDiscovery() {
 	fmt.Printf("Registered %v scheme\n", jsonDiscovery.Scheme())
 }
 
-type resolveFilters struct {
-	gate_type string
-	az_id     string
-}
+type hostFilters = map[string]string
 
 // exampleResolver is a
 // Resolver(https://godoc.org/google.golang.org/grpc/resolver#Resolver).
-type resolveJSONGateConfig struct {
-	target          resolver.Target
-	cc              resolver.ClientConn
-	jsonPath        string
-	ticker          *time.Ticker
-	rand            *rand.Rand // safe for concurrent use.
-	num_connections int
-	filters         resolveFilters
+type JSONGateConfigResolver struct {
+	target   resolver.Target
+	cc       resolver.ClientConn
+	jsonPath string
+	ticker   *time.Ticker
+	rand     *rand.Rand // safe for concurrent use.
+	filters  hostFilters
 }
 
-type discoverySlackAZ struct{}
-type discoverySlackType struct{}
+type matchesFilter struct{}
 
-func (r *resolveJSONGateConfig) loadConfig() (*[]resolver.Address, []byte, error) {
-	config := []DiscoveryHost{}
-	fmt.Printf("Loading config %v\n", r.jsonPath)
+func (r *JSONGateConfigResolver) loadConfig() (*[]resolver.Address, []byte, error) {
+	pairs := []map[string]interface{}{}
+	fmt.Printf("Loading config %v config for %v connections\n", r.jsonPath, *numConnectionsInt)
 
 	data, err := os.ReadFile(r.jsonPath)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	err = json.Unmarshal(data, &config)
+	err = json.Unmarshal(data, &pairs)
 	if err != nil {
 		fmt.Printf("parse err: %v\n", err)
 		return nil, nil, err
 	}
 
-	addrs := []resolver.Address{}
-	for _, s := range config {
-		az := attributes.New(discoverySlackAZ{}, s.AZId).WithValue(discoverySlackType{}, s.Type)
+	allAddrs := []resolver.Address{}
+	filteredAddrs := []resolver.Address{}
+	var addrs []resolver.Address
+	for _, pair := range pairs {
+		matchesAll := true
+		for k, v := range r.filters {
+			if pair[k] != v {
+				matchesAll = false
+			}
+		}
 
-		// Filter hosts to this gate type
-		if r.filters.gate_type != "" {
-			if r.filters.gate_type != s.Type {
-				continue
+		if matchesAll {
+			filteredAddrs = append(filteredAddrs, resolver.Address{
+				Addr:               fmt.Sprintf("%s:%s", pair["nebula_address"], pair["grpc"]),
+				BalancerAttributes: attributes.New(matchesFilter{}, "match"),
+			})
+		}
+
+		// Must filter by type
```

Comment:
This is totally fine for the Slack use case, but I'm not sure is something we'll want to always have in the code for when we upstream it.

More fundamentally, I feel like the code as-written is trying to shoehorn two different behaviors into the same `filters` list.

The `type` is, in fact, a filter, since if that's specified and there aren't enough hosts that match, we will fail to route.

The other field(s) are actually an affinity but we're of course willing to route across AZs if there aren't enough available. Also we don't _actually_ have any use case for more than one affinity field, and I'm having a hard time finding any reason for anyone upstream to really want that.

So, my suggestion is that we actually make this less generic, and rewrite this so that instead of the combined `filters` list, we change the signature of the resolver to something where we take in explicit configuration of something like:

```
type JSONGateConfigResolver struct {
        // If a field name is specified then filter the set of hosts to only those that match 
        // the given target, e.g. if targetTypeField is set to "type", then a route target of
        // vtgate://foo only matches hosts with "type: foo" in the host list
	targetTypeField string

        // If a field name is specified (e.g. "az_id"), then prefer routing to hosts that match
        // the affinity expressed in the target, e.g. if targetAffinityField is set to "az_id", 
        // then a target of vtgate://foo?az_id=1a will prefer hosts with "az_id: 1a" in the
        // host list, but may route to others if there are not enough candidates 
	targetAffinityField  hostFilters

	target   resolver.Target
	cc       resolver.ClientConn
	ticker   *time.Ticker
	rand     *rand.Rand // safe for concurrent use.

```

I think not only the comments but the more explicit separation between the two would make a lot of sense to me.